src/metabase/api/database.clj

```diff
@@ -404,54 +404,106 @@
      :limit     limit}))
 
-(defn- autocomplete-results [tables fields limit]
-  (let [tbl-count   (count tables)
-        fld-count   (count fields)
-        take-tables (min tbl-count (- limit (/ fld-count 2)))
-        take-fields (- limit take-tables)]
-    (concat (for [{table-name :name} (take take-tables tables)]
-              [table-name "Table"])
-            (for [{:keys [table_name base_type semantic_type name]} (take take-fields fields)]
-              [name (str table_name
-                         " "
-                         base_type
-                         (when semantic_type
-                           (str " " semantic_type)))]))))
-
-(defn- autocomplete-suggestions
-  "match-string is a string that will be used with ilike. The it will be lowercased by autocomplete-{tables,fields}. "
-  [db-id match-string]
-  (let [limit  50
-        tables (filter mi/can-read? (autocomplete-tables db-id match-string limit))
-        fields (readable-fields-only (autocomplete-fields db-id match-string limit))]
-    (autocomplete-results tables fields limit)))
+(defn- autocomplete-card-columns
+  "Returns a list of card column metadata given the `search-string` and a set of card ids.
+   The search-string is matched against column names in the result_metadata of the cards.
+  Each result contains all the metadata for the column, and the name of the card."
+  [db-id search-string limit match-type card-ids]
+  (when (seq card-ids)
+    (->> (db/select [Card :name :result_metadata :collection_id]
+           :%lower.result_metadata         [:like (match-search-string :substring search-string)]
+           :database_id                    db-id
+           {:where [:in :id card-ids]
+            :limit limit})
+         (filter mi/can-read?)
+         (mapcat (fn [{:keys [result_metadata name]}]
+                   (map (fn [column-metadata]
+                          (merge column-metadata
+                                 {:card_name name}))
+                        result_metadata)))
+         (filter (fn [column-metadata]
+                   (re-find (re-pattern (case match-type
+                                          :prefix    (str "(?i)^" search-string)
+                                          :substring (str "(?i)" search-string))) (:name column-metadata))))
+         (sort-by :name))))
+
+(defmulti format-autocomplete-result
+  "Format an autocomplete result for the client."
+  {:arglists '([result-data])}
+  (fn [[result-type _]]
+    result-type))
+
+(defmethod format-autocomplete-result :table
+  [[_ {:keys [name]}]]
+  [name "Table"])
+
+(defmethod format-autocomplete-result :field
+  [[_ {:keys [table_name base_type semantic_type name]}]]
+  [name (str table_name
+             " "
+             base_type
+             (when semantic_type
+               (str " " semantic_type)))])
+
+(defmethod format-autocomplete-result :card-column
+  [[_ {:keys [card_name base_type semantic_type name]}]]
+  [name (str card_name
```

Comment:
Maybe this is more of a production question:
We're using name as an implicit separator here, but the thing with `card name` is that it often has spaces and is quite long.

So currently if you have a question named "People in CA state",
And the suggestion will look like this:
`email People in CA state :type/email`

This looks a bit confusing to me, how about we use the `card-id` instead?
something like

`email question/#card-id :type/Email`

Reply:
Personally I would rather truncate a long name like `email   People in CA...   :type/Email`, but this is definitely something I would defer to product on.